### PR TITLE
Ppu adding Tile viewer debugger

### LINKED ISSUE #38: move the tile rendered window to it's dedicated debugger

### PR DESCRIPTION
**Pull request by Octopus773**

### CODE REVIEW

File: sources/Debugger/TileViewer/RAMTileRenderer.cpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+		this->_tileRenderer.setRam(ram);
+	}
+
+	void RAMTileRenderer::render()
+	{
+		int bufX = 0;
+		int bufY = 0;
+		int nbTilesDrawn = 0;
+		int resetX = bufX;
+		for (auto &i : buffer)
```
Comment (zoriya) [inline sources/Debugger/TileViewer/RAMTileRenderer.cpp]: You forgot the this->

File: sources/Debugger/TileViewer/TileViewer.cpp [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+	}
+
+	void TileViewer::_bppChangeUIHandler(int index)
+	{
+		switch (index) {
+		case 0: return this->setBpp(2);
+		case 1: return this->setBpp(4);
+		case 2: return this->setBpp(8);
+		default:
+			break;
```
Comment (zoriya) [inline sources/Debugger/TileViewer/TileViewer.cpp]: Please throw on unexpected cases.

File: sources/PPU/Background.cpp [hunk 3]
```diff
@@ -47,107 +52,50 @@ namespace ComSquare::PPU
...
-		graphicAddress = this->_tilesetAddress + (tileData.posY * NbTilePerRow * this->_bpp * TileBaseByteSize) + (tileData.posX * this->_bpp * TileBaseByteSize);
+
+		this->_tileRenderer.setPaletteIndex(tileData.palette);
+		for (int i = 0; i < this->_characterNbPixels.y; i += Tile::NbPixelsHeight) {
+			for (int j = 0; j < this->_characterNbPixels.x; j += Tile::NbPixelsWidth) {
+				graphicAddress = this->_tilesetAddress +
+				                 ((tileData.posY + tileOffset.y) * NbTilePerRow * this->_bpp * Tile::BaseByteSize) +
+				                 ((tileData.posX + tileOffset.x) * this->_bpp * Tile::BaseByteSize);
+				this->_tileRenderer.render(graphicAddress);
+				merge2DArray(this->_tileBuffer, this->_tileRenderer.buffer, {j, i});
```
Comment (zoriya) [inline sources/PPU/Background.cpp]: Put this in a static class or a namespace and call it explicitly. We don't really understand where they are now.

File: sources/PPU/PPU.hpp [hunk 4]
```diff
@@ -6,17 +6,18 @@
...
+#include "Ram/Ram.hpp"
+#include "Models/Vector2.hpp"
 #include "Background.hpp"
 #include "PPUUtils.hpp"
+#include "Debugger/TileViewer/RAMTileRenderer.hpp"
 
 #define FALLTHROUGH __attribute__((fallthrough));
 
-// TODO check if it usefull to have defines instead of constepxr
+// TODO check if it useful to have defines instead of constexpr
```
Comment (zoriya) [inline sources/PPU/PPU.hpp]: Spoiler: it is not. We should have constexpr

File: sources/PPU/PPU.hpp [hunk 5]
```diff
@@ -618,8 +620,10 @@ namespace ComSquare::PPU
 		//! @brief Render the Main and sub screen correctly
 		void renderMainAndSubScreen();
 		//! @brief Add a bg buffer to another buffer
-		template <std::size_t DEST_SIZE, std::size_t SRC_SIZE>
-		void add_buffer(std::array<std::array<uint32_t, DEST_SIZE>, DEST_SIZE> &bufferDest, std::array<std::array<uint32_t, SRC_SIZE>, SRC_SIZE> &bufferSrc);
+		template <std::size_t DEST_SIZE_X, std::size_t DEST_SIZE_Y, std::size_t SRC_SIZE_X, std::size_t SRC_SIZE_Y>
```
Comment (zoriya) [inline sources/PPU/PPU.hpp:623]: Shouldn't you put your templated function in the hpp?

File: sources/SNES.cpp [hunk 6]
```diff
@@ -180,4 +181,21 @@ namespace ComSquare
 				this->_registerViewer = std::make_unique<Debugger::RegisterViewer>(*this);
 		#endif
 	}
+
+	void SNES::disableTileViewerDebugging()
+	{
+		#ifdef DEBUGGER_ENABLED
+				this->_tileViewer = nullptr;
```
Comment (zoriya) [inline sources/SNES.cpp]: Indent (same for the function below)

File: sources/Renderer/QtRenderer/QtSFML.hpp [hunk 7]
```diff
@@ -37,27 +37,30 @@ namespace ComSquare::Renderer
...
+		~QtFullSFML() = default;
 	};
 
-	//! @brief A SFML renderer inside a QT window.
+	//! @brief A SFML renderer inside a QT widget.
 	class QtSFML : public IRenderer {
 	private:
 		//! @brief The main window that the app reside on.
-		QMainWindow _window;
+		QWidget *_window;
```
Comment (melefo) [inline sources/Renderer/QtRenderer/QtSFML.hpp:56]: Maybe put this raw pointer to a smart one?

**Review by melefo — APPROVED** [review]

File: sources/Renderer/QtRenderer/QtSFML.hpp (hunk 7 above)
Comment (zoriya) [reply sources/Renderer/QtRenderer/QtSFML.hpp:56]: This is a QWidget, Qt handles everything internally with news and deletes. We can't really do something for that.

File: sources/Debugger/TileViewer/RAMTileRenderer.cpp (hunk 1 above)
Comment (Octopus773) [reply sources/Debugger/TileViewer/RAMTileRenderer.cpp]: fixed

File: sources/Debugger/TileViewer/TileViewer.cpp (hunk 2 above)
Comment (Octopus773) [reply sources/Debugger/TileViewer/TileViewer.cpp]: fixed

Comment (Octopus773) [reply sources/Debugger/TileViewer/TileViewer.cpp]: fixed

File: sources/PPU/Background.cpp (hunk 3 above)
Comment (Octopus773) [reply sources/PPU/Background.cpp]: fixed

File: sources/PPU/PPU.hpp (hunk 5 above)
Comment (Octopus773) [reply sources/PPU/PPU.hpp:623]: fixed

File: sources/SNES.cpp (hunk 6 above)
Comment (Octopus773) [reply sources/SNES.cpp]: fixed

File: sources/Renderer/QtRenderer/QtSFML.hpp (hunk 7 above)
Comment (Octopus773) [reply sources/Renderer/QtRenderer/QtSFML.hpp:56]: idk honestly so i won't change it

File: sources/PPU/PPU.hpp (hunk 4 above)
Comment (Octopus773) [reply sources/PPU/PPU.hpp]: fixed

**Review by zoriya — APPROVED** [review]

**Pull request merged**